File: leetcode-practice/08-06-2026/Problem 1/solution.py
# Python (Brute Force) -- O(n^2)

class Solution:
    def twoSum(self, nums, target):

        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):

                if nums[i] + nums[j] == target:
                    return [i, j]
                

# Python (Hash Map) -- optimal solution O(n)

# Two Sum - Optimal Solution (O(n))

def two_sum(nums, target):

    hashmap = {}

    for i, num in enumerate(nums):

        complement = target - num

        if complement in hashmap:
            return [hashmap[complement], i]

        hashmap[num] = i

    return []


# Test Cases
print(two_sum([2, 7, 11, 15], 9))
print(two_sum([3, 2, 4], 6))
print(two_sum([3, 3], 6))

File: leetcode-practice/08-06-2026/Problem 1/solution.js
// JavaScript (Brute Force) -- time complexity O(n^2)


function twoSum1(nums, target) {
    for (let i = 0; i < nums.length; i++) {
        for (let j = i + 1; j < nums.length; j++) {
            if (nums[i] + nums[j] === target) {
                return [i, j];
            }
        }
    }
}


// JavaScript (Hash Map) -- optimal solution O(n)


// Two Sum - Optimal Solution (O(n))

function twoSum2(nums, target) {
    const map = new Map();

    for (let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];

        if (map.has(complement)) {
            return [map.get(complement), i];
        }

        map.set(nums[i], i);
    }

    return [];
}

// Test Cases
console.log(twoSum2([2, 7, 11, 15], 9)); // [0,1]
console.log(twoSum2([3, 2, 4], 6));      // [1,2]
console.log(twoSum2([3, 3], 6));         // [0,1]

